arvo/clay/data-types.md

```diff
@@ -309,342 +351,308 @@ the desks we know about on their ship.
-is not a complicated type; it is not difficult to work out the meaning.
-
-#### `++nori:clay`, repository action
-
-```hoon
-    ++  nori                                                ::  repository action
-              $%  [& q=soba]                                ::  delta
-                  [| p=@tas]                                ::  label
-              ==                                            ::
-```
-
-This describes a change that we are asking clay to make to the desk.
+This describes a change that we are asking Clay to make to the `desk`.
 There are two kinds of changes that may be made: we can modify files or
 we can apply a label to a commit.
 
 In the `|` case, we will simply label the current commit with the given
 label. In the `&` case, we will apply the given changes.
 
-#### `++soba:clay`, delta
+#### `$soba:clay`, delta
 
 ```hoon
-    ++  soba  ,[p=cart q=(list ,[p=path q=miso])]           ::  delta
+  +$  soba  (list [p=path q=miso])                      ::  delta
 ```
 
-This describes a set of changes to make to a desk. The `cart` is simply
-a pair of the old hash and the new hash of the desk. The list is a list
-of changes keyed by the file they're changing. Thus, the paths are paths
-to files to be changed while `miso` is a description of the change
-itself.
+This describes a `list` of changes to make to a `desk`. The `path`s are `path`s
+to files to be changed, and the corresponding `miso` value is a description of
+the change itself.
 
-#### `++miso:clay`, ankh delta
+#### `$miso:clay`, ankh delta
 
 ```hoon
-    ++  miso                                                ::  ankh delta
-              $%  [%del p=*]                                ::  delete
-                  [%ins p=*]                                ::  insert
-                  [%mut p=udon]                             ::  mutate
-              ==                                            ::
+  +$  miso                                              ::  ankh delta
+    $%  [%del ~]                                        ::  delete
+        [%ins p=cage]                                   ::  insert
+        [%dif p=cage]                                   ::  mutate from diff
+        [%mut p=cage]                                   ::  mutate from raw
+    ==                                                  ::
 ```
 
-There are three kinds of changes that may be made to a node in a desk.
-We can insert a file, in which case `p` is the contents of the new file.
-We can delete a file, in which case `p` is the contents of the old file.
-Finally, we can mutate that file, in which case the `udon` describes the
-changes we are applying to the file.
+There are four kinds of changes that may be made to a node in a `desk`.
 
-#### `++mizu:clay`, merged state
+`%del` deletes the node.
 
-```hoon
-    ++  mizu  ,[p=@u q=(map ,@ud tako) r=rang]              ::  new state
-```
+`%ins` inserts a file given by `p`.
+
+`%dif` is currently unimplemented.
```

Comment:
@philipcmonk is this right?

Reply:
Apparently yes

Reply:
I'm a bit confused as to what this really means. Marks implement diffs of files, but if diffs aren't implemented for ankhs (which I understand to be both, then what datatype has a notion of diff that marks implement? arch? Not sure that my question totally makes sense, but my initial impression of this (in the PR message) was that it somehow meant that Clay didn't do diffs at all, which definitely seemed wrong and must mean that my understanding of what ankhs are for is wrong.

Reply:
I guess this just means that there aren't diffs for changes in directory structure?

Reply:
`%dif` in a soba/miso is a way of specifying the new version of a file by just providing a diff on the old version, which would be implemented using `+pact` in the mark.  However, we still use `+diff` and `+pact` in marks for individual files for a few reasons.  One is that the object store can store blobs (files, as `+blob:clay`) directly or as a diff on another blob, which may be more efficient.  The other is that to perform a 3-way merge, you need to get the diff of each branch against the mergebase, and then when you merge those two diffs you get a 3rd diff, which you can `+pact` on the mergebase to get the new file.

> I guess this just means that there aren't diffs for changes in directory structure?

Right, we don't have a particular concept of changes in a directory structure.  We could define one in terms of additions and deletions, or we could encode renames as well.  Both 3-way merge and ted/diff implicitly use the former for now.  Git tries to derive renames heuristically.

Reply:
https://github.com/urbit/docs/pull/1055/commits/2e2d5757a653996d284afb35041d812710884af7